### PR TITLE
Add typing to lightning.tuner

### PR DESCRIPTION
**Pull request by aniketmaurya**
## What does this PR do?

Part of #7037.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/tuner/auto_gpu_select.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
 
-def pick_multiple_gpus(nb):
+def pick_multiple_gpus(nb: int):
```
Comment (justusschock) [inline pytorch_lightning/tuner/auto_gpu_select.py:19]: there is a returntype missing here

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 2]
```diff
@@ -13,8 +13,10 @@
 # limitations under the License
 import logging
 import os
-from typing import Optional, Tuple
+from typing import Optional, Tuple, TYPE_CHECKING
 
+if TYPE_CHECKING:
+    from pytorch_lightning import Trainer
```
Comment (justusschock) [inline pytorch_lightning/tuner/batch_size_scaling.py:19]: this won't work. Use `import pytorch_lightning as pl` instead and annotate with `'pl.Trainer'`

File: pytorch_lightning/tuner/tuning.py [hunk 3]
```diff
@@ -26,27 +28,28 @@
...
+    ) -> None:
         self.trainer.model_connector.copy_trainer_model_properties(model)
         # setup data, etc...
         self.trainer.train_loop.setup_fit(model, train_dataloader, val_dataloaders, datamodule)
         # hook
         self.trainer.data_connector.prepare_data(model)
 
-    def tune(self, model, train_dataloader, val_dataloaders, datamodule):
+    def tune(self, model: LightningModule, train_dataloader: Optional[DataLoader],
+             val_dataloaders: Optional[DataLoader], datamodule: Optional[LightningModule]) -> None:
```
Comment (justusschock) [inline pytorch_lightning/tuner/tuning.py:52]:
```suggestion
             val_dataloaders: Optional[DataLoader], datamodule: Optional[LightningDataModule]) -> None:
```

**Review by justusschock — CHANGES_REQUESTED** [review]

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 2 above)
Comment (aniketmaurya) [reply pytorch_lightning/tuner/batch_size_scaling.py]: @justusschock can you please mention the reason why it won't work?

Comment (justusschock) [reply pytorch_lightning/tuner/batch_size_scaling.py]: sphinx cannot deal with the forward references you provide here as a result from the optional import

File: pytorch_lightning/tuner/tuning.py [hunk 4]
```diff
@@ -12,10 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List, Optional, Union
+from typing import List, Optional, Union, TYPE_CHECKING
 
 from torch.utils.data import DataLoader
 
+if TYPE_CHECKING:
+    import pytorch_lightning as pl
```
Comment (justusschock) [inline pytorch_lightning/tuner/tuning.py:20]: this has to be imported without the `if TYPE_CHECKING` guard

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 5]
```diff
@@ -13,8 +13,10 @@
 # limitations under the License
 import logging
 import os
-from typing import Optional, Tuple
+from typing import Optional, Tuple, TYPE_CHECKING
 
+if TYPE_CHECKING:
+    import pytorch_lightning as pl
```
Comment (justusschock) [inline pytorch_lightning/tuner/batch_size_scaling.py:19]: this has to be imported without the `if TYPE_CHECKING` guard

File: pytorch_lightning/tuner/lr_finder.py [hunk 6]
```diff
@@ -311,7 +312,7 @@ def func():
 
         return func
 
-    def plot(self, suggest: bool = False, show: bool = False):
+    def plot(self, suggest: bool = False, show: bool = False) -> 'plt.Figure':
```
Comment (aniketmaurya) [inline pytorch_lightning/tuner/lr_finder.py:315]: 'plt.Figure' is causing PEP8 to fail. `./pytorch_lightning/tuner/lr_finder.py:315: [F821] undefined name 'plt'`
What should be done here? @justusschock

Comment (justusschock) [reply pytorch_lightning/tuner/lr_finder.py:315]: is pyplot imported there?

Comment (aniketmaurya) [reply pytorch_lightning/tuner/lr_finder.py:315]: pyplot is imported inside the function

**Comment by Borda** [comment]
@aniketmaurya any update here? can we resolve conflicts? :]

**Comment by aniketmaurya** [comment]
> @aniketmaurya any update here? can we resolve conflicts? :]

I'll resolve the conflicts in few hours.

**Comment by aniketmaurya** [comment]
But mistake closed the issue 😅

File: pytorch_lightning/tuner/lr_finder.py [hunk 7]
```diff
@@ -60,6 +63,174 @@ def _determine_lr_attr_name(trainer: 'pl.Trainer', model: 'pl.LightningModule')
     )
 
 
+def lr_find(
```
Comment (awaelchli) [inline pytorch_lightning/tuner/lr_finder.py:66]: could you move it to the old place. easier for us to review the changes :)

Comment (aniketmaurya) [reply pytorch_lightning/tuner/lr_finder.py]: done @awaelchli

File: pytorch_lightning/tuner/lr_finder.py [hunk 8]
```diff
@@ -346,7 +346,10 @@ def on_batch_start(self, trainer, pl_module):
 
         self.lrs.append(trainer.lr_schedulers[0]['scheduler'].lr[0])
 
-    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
+    def on_train_batch_end(
+        self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', outputs, batch, batch_idx: Optional[int],
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/tuner/lr_finder.py:350]:
```
outputs: Optional[Union[tensor, Dict[str, Any]]
batch: Any
```
outputs is what the user can return from training_step and it can be None (signaling skipping a step) a tensor (just the loss) or a dict with multiple items and at least a key with the name loss.

**Comment by Borda** [comment]
> Should I annotate method variables also @Borda ?

yes if possible and if not pls add inline annotation skip the line from checking...
